File: hash/go/imagehash/internal/bitpack/bitpack.go
// Package bitpack converts between boolean[][] hash arrays and lowercase hex strings.
// Row-major flatten, MSB-first per nibble, zero-padded to ceil(M*N / 4) chars.
package bitpack

import (
	"fmt"
	"math/big"
	"strings"
)

func Pack(bits [][]bool) string {
	h := len(bits)
	if h == 0 || len(bits[0]) == 0 {
		return ""
	}
	w := len(bits[0])
	total := h * w
	var sb strings.Builder
	sb.Grow(total)
	for y := 0; y < h; y++ {
		for x := 0; x < w; x++ {
			if bits[y][x] {
				sb.WriteByte('1')
			} else {
				sb.WriteByte('0')
			}
		}
	}
	width := (total + 3) / 4
	v, ok := new(big.Int).SetString(sb.String(), 2)
	if !ok {
		return ""
	}
	hex := v.Text(16)
	if len(hex) < width {
		return strings.Repeat("0", width-len(hex)) + hex
	}
	return hex
}

func UnpackSquare(hex string) ([][]bool, error) {
	bits, err := hexToBits(hex)
	if err != nil {
		return nil, err
	}
	totalBits := len(bits)
	n := 0
	for n*n < totalBits {
		n++
	}
	if n*n != totalBits {
		return nil, fmt.Errorf("hex length %d (%d bits) is not a square shape", len(hex), totalBits)
	}
	out := make([][]bool, n)
	idx := 0
	for y := 0; y < n; y++ {
		out[y] = make([]bool, n)
		for x := 0; x < n; x++ {
			out[y][x] = bits[idx]
			idx++
		}
	}
	return out, nil
}

func UnpackFlat(hex string, secondAxis int) ([][]bool, error) {
	bits, err := hexToBits(hex)
	if err != nil {
		return nil, err
	}
	totalBits := 14 * secondAxis
	if len(bits) < totalBits {
		return nil, fmt.Errorf("hex too short for 14x%d shape: %d bits", secondAxis, len(bits))
	}
	out := make([][]bool, 14)
	idx := len(bits) - totalBits
	for y := 0; y < 14; y++ {
		out[y] = make([]bool, secondAxis)
		for x := 0; x < secondAxis; x++ {
			out[y][x] = bits[idx]
			idx++
		}
	}
	return out, nil
}

func hexToBits(hex string) ([]bool, error) {
	if hex == "" {
		return nil, fmt.Errorf("empty hex")
	}
	for _, c := range hex {
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
			return nil, fmt.Errorf("invalid hex character %q in %q", c, hex)
		}
	}
	v, ok := new(big.Int).SetString(hex, 16)
	if !ok {
		return nil, fmt.Errorf("could not parse hex %q", hex)
	}
	totalBits := len(hex) * 4
	bits := make([]bool, totalBits)
	for i := 0; i < totalBits; i++ {
		bits[totalBits-1-i] = v.Bit(i) == 1
	}
	return bits, nil
}

File: hash/java/src/main/java/io/github/wmetcalf/rosettasquint/hash/internal/BitPack.java
package io.github.wmetcalf.rosettasquint.hash.internal;

import java.math.BigInteger;

/** Boolean array ↔ hex string conversion. Row-major MSB-first. */
public final class BitPack {

    public static String pack(boolean[][] bits) {
        int h = bits.length;
        int w = bits[0].length;
        int total = h * w;
        StringBuilder bs = new StringBuilder(total);
        for (int y = 0; y < h; y++)
            for (int x = 0; x < w; x++)
                bs.append(bits[y][x] ? '1' : '0');
        BigInteger value = new BigInteger(bs.toString(), 2);
        int width = (total + 3) / 4;
        String hex = value.toString(16);
        if (hex.length() < width) {
            StringBuilder padded = new StringBuilder(width);
            for (int i = hex.length(); i < width; i++) padded.append('0');
            padded.append(hex);
            return padded.toString();
        }
        return hex;
    }

    public static boolean[][] unpackSquare(String hex) {
        int totalBits = hex.length() * 4;
        int n = (int) Math.round(Math.sqrt(totalBits));
        if (n * n != totalBits) {
            throw new IllegalArgumentException("Hex length " + hex.length() + " doesn't correspond to a square shape (total bits=" + totalBits + ", sqrt=" + Math.sqrt(totalBits) + ")");
        }
        boolean[] flat = hexToBits(hex, totalBits);
        boolean[][] out = new boolean[n][n];
        int idx = 0;
        for (int y = 0; y < n; y++)
            for (int x = 0; x < n; x++)
                out[y][x] = flat[idx++];
        return out;
    }

    public static boolean[][] unpackFlat(String hex, int secondAxis) {
        int totalBits = 14 * secondAxis;
        boolean[] flat = hexToBits(hex, totalBits);
        boolean[][] out = new boolean[14][secondAxis];
        int idx = 0;
        for (int y = 0; y < 14; y++)
            for (int x = 0; x < secondAxis; x++)
                out[y][x] = flat[idx++];
        return out;
    }

    private static boolean[] hexToBits(String hex, int totalBits) {
        if (!hex.matches("[0-9a-f]+"))
            throw new IllegalArgumentException("Invalid hex: " + hex);
        BigInteger value = new BigInteger(hex, 16);
        boolean[] bits = new boolean[totalBits];
        for (int i = 0; i < totalBits; i++) {
            bits[totalBits - 1 - i] = value.testBit(i);
        }
        return bits;
    }

    private BitPack() {}
}
